vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,363 @@
+
+}
+
+WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString)
+  : WinRTWebSocketResource(Uri{ Utf8ToUtf16(urlString) })
+{
+}
+
+WinRTWebSocketResource::~WinRTWebSocketResource() /*override*/
+{
+  Stop();
+}
+
+#pragma region Private members
+
+fire_and_forget WinRTWebSocketResource::PerformConnect()
+{
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    co_await m_socket.ConnectAsync(m_uri);
+
+    if (m_connectHandler)
+    {
+      m_connectHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    hr = e.code();
+  }
+
+  if (!SUCCEEDED(hr))
+  {
+    hresult_error e{ hr };
+    //OutputDebugString("WebSocket.connect failed (0x%8X) %ls\n", e);
+
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Connection });
+    }
+  }
+
+  // m_connectPerformed.Set();
+  m_connectPerformed.set_value();
+  m_connectRequested = false;
+}
+
+fire_and_forget WinRTWebSocketResource::PerformPing()
+{
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    if (m_connectRequested)
+    {
+      m_connectPerformed.get_future().wait();
+    }
+
+    m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+    string s{};
+    winrt::array_view<const uint8_t> arr(
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()),
+      CheckedReinterpretCast<const uint8_t*>(s.c_str()) + s.length());
+    m_writer.WriteBytes(arr);
+
+    co_await m_writer.StoreAsync();
+
+    if (m_pingHandler)
+    {
+      m_pingHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    hr = e.code();
+  }
+
+  if (!SUCCEEDED(hr))
+  {
+    if (m_errorHandler)
+    {
+      m_errorHandler({ Utf16ToUtf8(hresult_error(hr).message()), ErrorType::Ping });
+    }
+  }
+}
+
+fire_and_forget WinRTWebSocketResource::PerformWrite()
+{
+  if (m_writeQueue.empty())
+  {
+    co_return;
+  }
+
+  hresult hr = S_OK;
+  try
+  {
+    co_await resume_background();
+
+    if (m_connectRequested)
+    {
+      m_connectPerformed.get_future().wait();
+    }
+
+    string message;
+    bool isBinary;
+    size_t length;
+    std::tie(message, isBinary) = m_writeQueue.front();
+    m_writeQueue.pop();
+
+    if (isBinary)
+    {
+      m_socket.Control().MessageType(SocketMessageType::Binary);
+
+      auto buffer = CryptographicBuffer::DecodeFromBase64String(Utf8ToUtf16(std::move(message)));
+      length = buffer.Length();
+      m_writer.WriteBuffer(buffer);
+    }
+    else
+    {
+      m_socket.Control().MessageType(SocketMessageType::Utf8);
+
+      length = message.size();
+      winrt::array_view<const uint8_t> arr(
+        CheckedReinterpretCast<const uint8_t*>(message.c_str()),
+        CheckedReinterpretCast<const uint8_t*>(message.c_str()) + message.length());
+      m_writer.WriteBytes(arr);
+    }
+
+    co_await m_writer.StoreAsync();
+
+    if (m_writeHandler)
+    {
+      m_writeHandler(length);
+    }
+
+    if (!m_writeQueue.empty())
+    {
+      PerformWrite();
```

Comment:
This doesn't feel right.

Methods will queue up items in the queue using `PerformWrite()`, but also seem to spawn a consumer immediately. I don't see any logic here to bail if something is already processing the queue. I.e. currently it seems like multiple things could be already pulling from the queue at once.

Reply:
Correct.
This currently lacks the proper sequencing logic to avoid concurrent accesses to critical section wile writing.

This is the last major issue being worked on.
Note, the previous implementation relied on `boost::asio::io_context`, which we may not be able to use here.